src/NHibernate/Action/CollectionAction.cs

```diff
@@ -58,8 +58,8 @@ protected internal object Key
 				finalKey = key;
 				if (key is DelayedPostInsertIdentifier)
 				{
-					// need to look it up from the persistence-context
-					finalKey = session.PersistenceContext.GetEntry(collection.Owner).Id;
+					// need to look it up
+					finalKey = persister.CollectionType.GetKeyOfOwner(collection.Owner, session);
```

Comment:
There was already some fallback in collection action for resolving a `DelayedPostInsertIdentifier`, but it was invalid: the collection may be something else than the owner Id. It may be a `many-to-one` through `property-ref`. This was likely not having any consequences since `Persist`, prior to this PR, was anyway not using `DelayedPostInsertIdentifier`.

The `CollectionType` will do the previous code if there is no `property-ref`, otherwise it will do the appropriate code for getting the actual key.